crates/astria-proto/src/native/sequencer/v1alpha1.rs

```diff
@@ -1251,6 +1266,1158 @@ impl Display for IncorrectAddressLength {
+        }
+
+        let Self {
+            header,
+            rollup_transactions,
+            rollup_transactions_proof,
+            rollup_ids_proof,
+            ..
+        } = self;
+        raw::SequencerBlock {
+            header: Some(header.into()),
+            rollup_transactions: rollup_transactions
+                .into_iter()
+                .map(tuple_to_rollup_txs)
+                .collect(),
+            rollup_transactions_proof: Some(rollup_transactions_proof.into_raw()),
+            rollup_ids_proof: Some(rollup_ids_proof.into_raw()),
+        }
+    }
+
+    #[must_use]
+    pub fn into_unchecked(self) -> UncheckedSequencerBlock {
+        let Self {
+            header,
+            rollup_transactions,
+            rollup_transactions_proof,
+            rollup_ids_proof,
+            ..
+        } = self;
+        UncheckedSequencerBlock {
+            header,
+            rollup_transactions,
+            rollup_transactions_proof,
+            rollup_ids_proof,
+        }
+    }
+
+    /// Turn the sequencer block into a [`CelestiaSequencerBlob`] and its associated list of
+    /// [`CelestiaRollupBlob`]s.
+    #[must_use]
+    #[allow(clippy::missing_panics_doc)] // the proofs are guaranteed to exist; revisit if refactoring
+    pub fn to_celestia_blobs(&self) -> (CelestiaSequencerBlob, Vec<CelestiaRollupBlob>) {
+        let SequencerBlock {
+            block_hash,
+            header,
+            rollup_transactions,
+            rollup_transactions_proof,
+            rollup_ids_proof,
+        } = self;
+
+        let tree = derive_merkle_tree_from_rollup_txs(rollup_transactions);
+
+        let head = CelestiaSequencerBlob {
+            block_hash: *block_hash,
+            header: header.clone(),
+            rollup_ids: rollup_transactions.keys().copied().collect(),
+            rollup_transactions_root: tree.root(),
+            rollup_transactions_proof: rollup_transactions_proof.clone(),
+            rollup_ids_proof: rollup_ids_proof.clone(),
+        };
+
+        let mut tail = Vec::with_capacity(self.rollup_transactions.len());
+        for (i, (rollup_id, transactions)) in self.rollup_transactions.iter().enumerate() {
+            let proof = tree
+                .construct_proof(i)
+                .expect("the proof must exist because the tree was derived with the same leaf");
+            tail.push(CelestiaRollupBlob {
+                sequencer_block_hash: self.block_hash(),
+                rollup_id: *rollup_id,
+                transactions: transactions.clone(),
+                proof,
+            });
+        }
+        (head, tail)
+    }
+
+    /// Converts from a [`tendermint::Block`].
+    ///
+    /// # Errors
+    /// TODO(https://github.com/astriaorg/astria/issues/612)
+    #[allow(clippy::missing_panics_doc)] // the panic sources are checked before hand; revisit if refactoring
+    pub fn try_from_cometbft(block: tendermint::Block) -> Result<Self, SequencerBlockError> {
```

Comment:
This was moved from `astria-sequencer-types` and enriched by checking more invariants.